Fix LogConfig in log admin REST API Issue #3707

ArqamFarooqui110719 opened the pull request:
Fix LogConfigs in log admin REST API Issue #3707

fgalan commented:
Thank you for your contribution!

Basic functional test (.test file) covering the implemented functionality should be added (maybe this helps: https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/devel/cookbook.md#adding-a-functional-test-case).

In addition, note that some tests have failed in GitHub regression. They have to be fixed before merging this PR.

Finally, an entry to CHANGES_NEXT_RELEASE file regarding the functionality in this PR should be included.

ArqamFarooqui110719 commented:
Yes. I am writing new FTs and will also fix the existing failing FTs.

fgalan commented:
Please have also a look to documentation in https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/admin/management_api.md#log-and-trace-levels. I understand it should be modified to take into account the extra parameters in the API calls.

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -2,3 +2,4 @@
 - Fix: queue workers deadlock when MQTT broker was not responding to connection attemp
 - Fix: Ngsiv2-AttrsFormat header set to "normalized" in NGSI patching custom notification
 - Fix: Orion should return error and refuse to start when port is negative (3875)
+- Fix: logLineMaxSize and logLineMaxSize in the log admin REST API (3707)
```
Text-style:

```suggestion'
- Fix: Orion should return error and refuse to start when port is negative (#3875)
- Fix: logLineMaxSize and logLineMaxSize in the log admin REST API (#3707)
```

ArqamFarooqui110719 replied:
Fix in [18cb428](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/18cb428517740d4662bb8f7a325f4ca6e7692b8a)

ArqamFarooqui110719 commented:
> Please have also a look to documentation in https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/admin/management_api.md#log-and-trace-levels. I understand it should be modified to take into account the extra parameters in the API calls.

Updated the `management_api.md` document as per new API calls and added the extra parameter.

fgalan commented on doc/manuals/admin/management_api.md:
````diff
@@ -1,17 +1,19 @@
 ```
 
-To retrieve the log level:
+To retrieve the log Config:
````
```suggestion
To retrieve the log config:
```

ArqamFarooqui110719 replied:
Fixed in [2c97550](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/2c97550fc344e0c951be6811fba52d3472fb8c8a)

ArqamFarooqui110719 commented:
Hi @fgalan sir, Kindly review it and If the PR is OK, then please merge it.

fgalan commented:
In cli.md document we have:

> -   **-logLineMaxSize**. Log line maximum length (when exceeded Orion prints `LINE TOO LONG` as log trace). Minimum allowed value: 100 bytes. Default value: 32 KBytes.
> -   **-logInfoPayloadMaxSize**. For those log traces at INFO level that print request and/or response payloads, this is the maximum allowed size for those payloads. If the payload size is greater than this setting, then only the first `-logInfoPayloadMaxSize` bytes are included (and an ellipsis in the form of `(...)` is shown in trace). Default value: 5 KBytes.

I'd suggest to change them, so it reads:

> -   **-logLineMaxSize**. Log line maximum length (when exceeded Orion prints `LINE TOO LONG` as log trace). Minimum allowed value: 100 bytes. Default value: 32 KBytes. It can be changed after Orion startup with the [log admin REST API](management_api.md#log-and-trace-levels).
> -   **-logInfoPayloadMaxSize**. For those log traces at INFO level that print request and/or response payloads, this is the maximum allowed size for those payloads. If the payload size is greater than this setting, then only the first `-logInfoPayloadMaxSize` bytes are included (and an ellipsis in the form of `(...)` is shown in trace). Default value: 5 KBytes. It can be changed after Orion startup with the [log admin REST API](management_api.md#log-and-trace-levels).

fgalan commented on src/lib/serviceRoutinesV2/logLevelTreat.cpp:
```diff
@@ -55,11 +55,44 @@ std::string changeLogLevel
+    {
+      ciP->httpStatusCode = SccBadRequest;
+      alarmMgr.badInput(clientIp, "invalid logInfoPayloadMaxSize in URI param", payloadSize);
+      return "{\"error\":\"invalid logInfoPayloadMaxSize,logPayload size should be > 0\"}";
```
```suggestion
      return "{\"error\":\"invalid logInfoPayloadMaxSize, logPayload size should be > 0\"}";
```

(Changes in .test would be also required)

ArqamFarooqui110719 replied:
Fixed in [46582d5](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/46582d5125caad31b3df9116faa6797957652b95)

ArqamFarooqui110719 commented:
> In cli.md document we have:
> 
> > * **-logLineMaxSize**. Log line maximum length (when exceeded Orion prints `LINE TOO LONG` as log trace). Minimum allowed value: 100 bytes. Default value: 32 KBytes.
> > * **-logInfoPayloadMaxSize**. For those log traces at INFO level that print request and/or response payloads, this is the maximum allowed size for those payloads. If the payload size is greater than this setting, then only the first `-logInfoPayloadMaxSize` bytes are included (and an ellipsis in the form of `(...)` is shown in trace). Default value: 5 KBytes.
> 
> I'd suggest to change it, so it reads:
> 
> > * **-logLineMaxSize**. Log line maximum length (when exceeded Orion prints `LINE TOO LONG` as log trace). Minimum allowed value: 100 bytes. Default value: 32 KBytes. It can be changed after Orion startup with the [log admin REST API](management_api.md#log-and-trace-levels).
> > * **-logInfoPayloadMaxSize**. For those log traces at INFO level that print request and/or response payloads, this is the maximum allowed size for those payloads. If the payload size is greater than this setting, then only the first `-logInfoPayloadMaxSize` bytes are included (and an ellipsis in the form of `(...)` is shown in trace). Default value: 5 KBytes. It can be changed after Orion startup with the [log admin REST API](management_api.md#log-and-trace-levels).

updated cli.md doc in [46582d5](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/46582d5125caad31b3df9116faa6797957652b95)

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -1,4 +1,5 @@
 - Fix: Ngsiv2-AttrsFormat header set to "normalized" in NGSI patching custom notification
-- Fix: Orion should return error and refuse to start when port is negative (3875)
+- Fix: Orion should return error and refuse to start when port is negative (#3875)
+- Fix: logLineMaxSize and logLineMaxSize in the log admin REST API (#3707)
```
Little typo:

```suggestion
- Fix: logLineMaxSize and logInfoPayloadMaxSize in the log admin REST API (#3707)
```

ArqamFarooqui110719 replied:
Fixed in [24de8d9](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/24de8d9e544ab1ddba1af46cdb70ded49bb85e2e)

fgalan commented on src/lib/serviceRoutinesV2/logLevelTreat.cpp:
```diff
@@ -55,11 +55,44 @@ std::string changeLogLevel
+    {
+      ciP->httpStatusCode = SccBadRequest;
+      alarmMgr.badInput(clientIp, "invalid logLineMaxSize in URI param", logSize);
+      return "{\"error\":\"invalid logLineMaxSize, logLine size should be > 0\"}";
```
```suggestion
      return "{\"error\":\"invalid lineMaxSize, lineMaxSieze should be > 0\"}";
```

To align with the name of the field that provides this information in the PUT request.

ArqamFarooqui110719 replied:
Fixed in [24de8d9](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/24de8d9e544ab1ddba1af46cdb70ded49bb85e2e)

fgalan commented on test/functionalTest/cases/3707_get_or_change_log_config/get_or_change_log_config.test:
```diff
@@ -0,0 +1,488 @@
+# 17. GET log config, see lineMaxSize=32000
+# 18. PUT /admin/log?lineMaxSize=400000 and see it work
+# 19. GET log config, see lineMaxSize=400000
+# 20. PUT /admin/log?lineMaxSize=-30000 and see error
```
I'd suggest to add a couple of cases with non integer number. For instance:

```
xx. PUT /admin/log?lineMaxSize=3.1416 and see error
xx. PUT /admin/log?lineMaxSize=nonumer and see error
```

ArqamFarooqui110719 replied:
Fixed+Added cases in [24de8d9](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/24de8d9e544ab1ddba1af46cdb70ded49bb85e2e)

fgalan commented on test/functionalTest/cases/3707_get_or_change_log_config/get_or_change_log_config.test:
```diff
@@ -0,0 +1,488 @@
+# 11. GET log config, see infoPayloadMaxSize=5000
+# 12. PUT /admin/log?infoPayloadMaxSize=100000 and see it work
+# 13. GET log config, see infoPayloadMaxSize=100000
+# 14. PUT /admin/log?infoPayloadMaxSize=-2000 and see error
```
I'd suggest to add a couple of cases with non integer number. For instance:

```
xx. PUT /admin/log?infoPayloadMaxSize=3.1416 and see error
xx. PUT /admin/log?infoPayloadMaxSize=nonumer and see error
```

ArqamFarooqui110719 replied:
Fixed. Added cases in [24de8d9](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/24de8d9e544ab1ddba1af46cdb70ded49bb85e2e)

ArqamFarooqui110719 commented on src/lib/serviceRoutinesV2/logLevelTreat.cpp:
```diff
@@ -55,11 +55,44 @@ std::string changeLogLevel
+    {
+      ciP->httpStatusCode = SccBadRequest;
+      alarmMgr.badInput(clientIp, "invalid logInfoPayloadMaxSize in URI param", payloadSize);
+      return "{\"error\":\"invalid logInfoPayloadMaxSize, logPayload size should be > 0\"}";
```
Fixed in [24de8d9](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/24de8d9e544ab1ddba1af46cdb70ded49bb85e2e)

ArqamFarooqui110719 commented on doc/manuals/admin/cli.md:
```diff
@@ -174,8 +174,8 @@ The list of available options is the following:
--   **-logLineMaxSize**. Log line maximum length (when exceeded Orion prints `LINE TOO LONG` as log trace). Minimum allowed value: 100 bytes. Default value: 32 KBytes.
--   **-logInfoPayloadMaxSize**. For those log traces at INFO level that print request and/or response payloads, this is the maximum allowed size for those payloads. If the payload size is greater than this setting, then only the first `-logInfoPayloadMaxSize` bytes are included (and an ellipsis in the form of `(...)` is shown in trace). Default value: 5 KBytes.
+-   **-logLineMaxSize**. Log line maximum length (when exceeded Orion prints `LINE TOO LONG` as log trace). Minimum allowed value: 100 bytes. Default value: 32 KBytes. It can be changed after Orion startup with the [log admin REST API](management_api.md#log-configs-and-trace-levels).
+-   **-logInfoPayloadMaxSize**. For those log traces at INFO level that print request and/or response payloads, this is the maximum allowed size for those payloads. If the payload size is greater than this setting, then only the first `-logInfoPayloadMaxSize` bytes are included (and an ellipsis in the form of `(...)` is shown in trace). Default value: 5 KBytes. It can be changed after Orion start ...
```
Fixed in [24de8d9](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/24de8d9e544ab1ddba1af46cdb70ded49bb85e2e)

ArqamFarooqui110719 commented on doc/manuals/admin/cli.md:
```diff
@@ -174,8 +174,8 @@ The list of available options is the following:
 -   **-logForHumans**. To make the traces to standard out formated for humans (note that the traces in the log file are not affected)
--   **-logLineMaxSize**. Log line maximum length (when exceeded Orion prints `LINE TOO LONG` as log trace). Minimum allowed value: 100 bytes. Default value: 32 KBytes.
--   **-logInfoPayloadMaxSize**. For those log traces at INFO level that print request and/or response payloads, this is the maximum allowed size for those payloads. If the payload size is greater than this setting, then only the first `-logInfoPayloadMaxSize` bytes are included (and an ellipsis in the form of `(...)` is shown in trace). Default value: 5 KBytes.
+-   **-logLineMaxSize**. Log line maximum length (when exceeded Orion prints `LINE TOO LONG` as log trace). Minimum allowed value: 100 bytes. Default value: 32 KBytes. It can be changed after Orion startup with the [log admin REST API](management_api.md#log-configs-and-trace-levels).
```
Fixed in [24de8d9](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/24de8d9e544ab1ddba1af46cdb70ded49bb85e2e)

fgalan commented on test/functionalTest/cases/3707_get_or_change_log_config/get_or_change_log_config.test:
```diff
@@ -0,0 +1,758 @@
+Date: REGEX(.*)
+
+{
+    "infoPayloadMaxSize": "5000",
+    "level": "INFO",
+    "lineMaxSize": "32768"
```
I thinks it would be better:

```suggestion
    "infoPayloadMaxSize": 5000,
    "level": "INFO",
    "lineMaxSize": 32768
```

I mean, lineMaxSize and infoPayloadMaxSize as JSON numbers instead of JSON string meaning numbers

(Sorry for not detecting this in previous review rounds...)

ArqamFarooqui110719 replied:
fixed in [b72f2f3](https://github.com/telefonicaid/fiware-orion/pull/4289/commits/b72f2f32bebc08ed52276a61e8a43a851a78faed)

fgalan reviewed (approved):
LGTM

Thanks for the contribution!

fgalan commented:
@fisuda this PR does some modification to .md files. It would be great if you could sync in the Japanese translation. Thanks in advance!

fisuda commented:
I sent the PR #4292. Thanks.
